lilypond: add openlilylib-fonts and lilypond-with-fonts

yurrriq opened the pull request:
###### Motivation for this change

I want to use some custom music fonts, which LilyPond supports but needs for them to be in a particular location.

###### Notes

I feel like the code I've written is a bit sloppy and I'm very open to suggestions.

Example:

```nix
{
  nixpkgs.config.packageOverrides = pkgs: rec {
    my-lilypond = pkgs.lilypond-with-fonts(with pkgs.openlilylib-fonts; [
      improviso
      lilyjazz
    ]);
}
```

*Edit*:

```nix
{
    my-lilypond = pkgs.lilypond-with-fonts.override {
      fonts = with pkgs.openlilylib-fonts; [ improviso lilyjazz ];
    };
}
```




###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

yurrriq committed:
lilypond: add openlilylib-fonts and lilypond-with-fonts

yurrriq commented on pkgs/misc/lilypond/fonts.nix:
```diff
@@ -0,0 +1,122 @@
+in
+
+{
+  beethoven = olpFont {
```
I assembled this mostly manually. It should probably be scripted.

Mic92 replied:
Probably ok for the moment.

yurrriq commented on pkgs/misc/lilypond/with-fonts.nix:
```diff
@@ -0,0 +1,28 @@
+      install -m755 -Dt $out/bin ${lilypond}/bin/*
+
+      for p in $out/bin/*; do
+        substituteInPlace $p --replace "exec -a ${lilypond}" "exec -a $out"
```
This works, but feels very hackish. The `lilypond` binary seems to look for fonts in `../share/lilypond/${lilypond.version}/fonts` so simply symlinking the original, doesn't work.

Mic92 replied:
I currently do not see a different solution.

yurrriq commented:
/cc @puffnfresh @Mic92

Mic92 commented on pkgs/misc/lilypond/with-fonts.nix:
```diff
@@ -0,0 +1,28 @@
+{ stdenv, lilypond, lndir }: fonts: stdenv.mkDerivation {
```
Could be default value to just use all fonts available in nixpkgs? Some people just want to get something working without having to override a package.

yurrriq replied:
Sure. Is there a slick way to do that besides writing them out manually?

yurrriq replied:
Also, I'm not sure how to write lambdas with optional arguments in Nix..

yurrriq committed:
lilypond-with-fonts: use all fonts by default

Add a list of all fonts to openlilylib-fonts and use that as the default
fonts argument to lilypond-with-fonts.

Mic92 commented:
Thanks!
